public/app/features/alerting/unified/components/expressions/Expression.tsx

```diff
@@ -60,15 +60,16 @@ export const Expression: FC<ExpressionProps> = ({
   const isLoading = data && Object.values(data).some((d) => Boolean(d) && d.state === LoadingState.Loading);
   const hasResults = Array.isArray(data?.series) && !isLoading;
   const series = data?.series ?? [];
-  const seriesCount = series.length;
 
   const alertCondition = isAlertCondition ?? false;
-
+  const noDataSeries = series.filter((serie) => !Boolean(getSeriesValue(serie))).length;
   const groupedByState = {
-    [PromAlertingRuleState.Firing]: series.filter((serie) => getSeriesValue(serie) !== 0),
+    [PromAlertingRuleState.Firing]: series.filter((serie) => Boolean(getSeriesValue(serie))),
```

Comment:
How is this different from `getSeriesValue(serie) !== 0`?

Reply:
with `getSeriesValue(serie) !== 0` we were filtering values including undefined. With `Boolean(getSeriesValue(serie))` we filter only values with an actual value (not empty, undefined and null are filtered)

Reply:
We could use `compact` from lodash to remove undefined and null values from the array.
I think Vika made a good point here, could we add a comment maybe so it's clear what needs to be done here?
Shouldn't we apply the same logic for `PromAlertingRuleState.Inactive`?

Reply:
Maybe we could add a helper function that returns `groupedByState` and add a few tests to it covering these cases?

Reply:
> Shouldn't we apply the same logic for PromAlertingRuleState.Inactive?
I would say no, because this filter is specifically filtering the series that have a zero value.

Reply:
> I think Vika made a good point here, could we add a comment maybe so it's clear what needs to be done here?

💯

Reply:
done!

Reply:
> We could use compact from lodash to remove undefined and null values from the array.

yes, you're right, but I think in this case the code is more clear using the same Array function filter for both cases.